csrc/util_class.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// JNI methods needed by the Java Utils class rather than generic utilities needed by our code.
+
+#include "generated-headers.h"
+
+#include "util.h"
+#include "env.h"
+#include "keyutils.h"
+
+using namespace AmazonCorrettoCryptoProvider;
+
+extern "C" {
+
+/*
+ * Class:     com_amazon_corretto_crypto_provider_Utils
+ * Method:    getNativeBufferOffset
+ * Signature: (Ljava/nio/ByteBuffer;Ljava/nio/ByteBuffer;)J
+ */
+
+JNIEXPORT jlong JNICALL Java_com_amazon_corretto_crypto_provider_Utils_getNativeBufferOffset
+  (JNIEnv *env, jclass, jobject bufA, jobject bufB)
+{
+    jlong no_overlap = (jlong)0x80000000;
```

Comment:
This is greater than [Integer.MAX_VALUE](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Integer.html#MAX_VALUE) (2^31 - 1).

Check out the documentation on the [Java side of this method](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/Utils.java#L35-L46). And the related [helper function](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/Utils.java#L48-L56) which calls it.

The intention of this method is if two native byte buffers overlap to return the offset between them. This offset value can be positive (if `bufB` is after `bufA`) or negative (for the reverse) or zero (if `bufA` and `bufB` start at the same place). So, how can we indicate no overlap if positive, negative, and zero values are all valid overlaps?

Because of how Java is defined, the length of a direct byte buffer is at most `Integer.MAX_VALUE` (2^31 - 1). This means that it is impossible for two buffers to have an offset greater than that (since that must be past the end of one of them). Thus, we can use any value who's absolute value is greater than 2^31 - 1 to indicate no overlap. And that is exactly what we are doing here.

Reply:
i see, thanks for the explanation.